A51-custom-backend-metrics.md

```diff
@@ -0,0 +1,387 @@
+subscribing to the data. When a policy subscribes, it will indicate the interval at which it would like to get updates. 
+If multiple policies subscribe with different intervals, we will use the minimum of those intervals. 
+Note that the ORCA protocol involves the client telling the server the interval at which 
+it wants reports when the call starts, so if the client needs to change its requested interval, 
+it will need to cancel and restart the call.
+A client may change its requested interval at any time; this is expected to be a fairly rare operation 
+(usually triggered by a config change for one of the subscribing LB policies),
+It is important that a load balancing policy is able to re-config the frequency
+and should not require heavy-weight operations like reconnecting to backends or recreating subchannels or LB policies.
+
+Each policy that has subscribed to the data will get a callback whenever a new message is sent to the client on the call.
+As with per-request data, the protobuf deserialization will happen only once.
+The data provided to the callback should be exposed to the LB policy using the same type as used for per-request data.
+The API for subscribing to out-of-band reports must be available to LB policies at any level of the routing hierarchy, 
+including both leaf LB policies (whose children are subchannels) and parent LB policies (whose children are other LB policies).
+Note that there may be multiple LB policies in the routing hierarchy that are interested in subscribing.
+For example, when using xDS, the `xds_cluster_impl` policy may need to see backend metric data to 
+generate load reports for each cluster, and the WRR policy might need to see the same data for 
+endpoint-picking within a locality.
+Reports are published to all the subscribers as-is with no aggregation; 
+each LB policy is responsible for performing whatever aggregation is appropriate for its use.
+Note that the LB policy API should be designed such that individual LB policies that do not know or 
+care about out-of-band reporting should not have to do anything special to support them. 
+For example, it should be possible for a non-leaf LB policy to subscribe to out-of-band reports from
+a child policy like round_robin, which does not know anything about out-of-band-reports and will not contain any
+code to handle them.
+
+When a subchannel first establishes a connection, if at least one LB policy is subscribing to out-of-band reporting, 
+the client will immediately start the `StreamCoreMetrics` call to the backend. 
+The client does not need to wait for this call to be established before it starts sending other calls on the subchannel.
+If the `StreamCoreMetrics` call fails with status `UNIMPLEMENTED`, the client will not retry the call, 
+and the subscribing LB polic(ies) will not receive any callbacks.
+However, the client will record a channel trace event indicating that this has happened. 
+It will also log a message at priority ERROR.
```

Comment:
What is this talking about? It seems likely inappropriate in Java.

Reply:
Why the `ERROR` level message traced/[logged](https://github.com/grpc/grpc-java/blob/881747a63df79a7baae252de86447c0997c76f69/xds/src/main/java/io/grpc/xds/OrcaOobUtil.java#L508) in subchannelLogger is inappropriate?

Reply:
This was intended to work the same way that we handle UNIMPLEMENTED for client-side health checking.  What does Java do in that case?  The original design for client-side health checking did say we should log at ERROR in this case, but maybe Java didn't implement that.

Reply:
1. Why is it an ERROR? Seems it should be a WARNING since the implementation continues
2. WARNING logs are generally bad, because if something is only a _concern_ but not a _problem_ then logging all the time causes noise and hides legitimate problems. This is the same mentality used for golang not to have compiler warnings. There are some times it happens, but it should be exceedingly rare and certainly not as part of normal operation as is the case here
3. Java should not be logging by default. If we _can't_ propagate an error somewhere, then we may need to log, but that does demonstrate a design issue. We are a library and application authors do not want to see log noise (I'm aware of some cases we are already inappropriately logging, but several of those should be cleaned up and some can't easily.)

Reply:
Looks like Java [does do an error for health check](https://github.com/grpc/grpc-java/blob/a5829107c3dd771a43cb2ef9c6eb967aa13b8aa3/services/src/main/java/io/grpc/protobuf/services/HealthCheckingLoadBalancerFactory.java#L419-L421).

Per subchannel logging can spam things quite a lot, so essentially they "never should happen." With health check the expectation was there was explicit configuration in the service config. Basically, if UNIMPLEMENTED, then the service config is incorrect or something else is busted. For these metrics, I don't think we have a signal for whether this information is required or optional. Since there can also be per-RPC metrics, the LB policy might actually just be expected to receive the information from a different source.

If we have the error here, we are saying that using the util implies the server _must_ support the OOB service and we will not support it being optional.

Reply:
My expectation is that LB policies will need to be explicitly configured to request OOB backend metric data, since that's going to be a minority case, and that no one will configure that unless they expect the data to be available.  So I think the same argument would apply here as in the client-side health checking case.

Reply:
I am leaning towards a warning level log:
If users configured the client LB policy to have metrics but the server has not be properly configured yet, or during deployment, part of the backend has not been running the latest configuration.
This is an LB miss-configuration issue, even be transient. Creating a new connection may be successful. The main traffic might still going fine but the routing is not what the user is desired. Also, we are having ERROR for channel panic. So this miss configuration issue might be fair to get a WARNING: it also tells which server has which problem. Seems sufficient.

Reply:
> If we have the error here, we are saying that using the util implies the server must support the OOB service and we will not support it being optional.

I'm fine if we make this expectation explicit, in the gRFC and API. We will try to handle the case gracefully, but LBs will probably break if the server doesn't support it.

Reply:
> > If we have the error here, we are saying that using the util implies the server must support the OOB service and we will not support it being optional.
> 
> I'm fine if we make this expectation explicit, in the gRFC and API. We will try to handle the case gracefully, but LBs will probably break if the server doesn't support it.

I think warning is better based on previous conversation tho.

Reply:
I think this should be error, for consistency with what we do in the health checking case.  I think the two cases are exactly the same semantically.